kernel/kernel-api/src/main/java/io/delta/kernel/internal/ScanImpl.java

```diff
@@ -190,6 +206,56 @@ public void close() throws IOException {
         };
     }
 
+    private Optional<Predicate> getDataSkippingPredicate() {
+        Optional<Predicate> dataPredicate = getDataFilters();
+        if (!dataPredicate.isPresent()) { // no data filter
+            return Optional.empty();
+        }
+        return DataSkippingUtils.constructDataFilters(dataPredicate.get(), metadata.getSchema());
+    }
+
+    private CloseableIterator<FilteredColumnarBatch> applyDataSkipping(
+        TableClient tableClient,
+        CloseableIterator<FilteredColumnarBatch> scanFileIter,
+        Predicate dataFilter) {
+        // Get the stats schema
+        // TODO prune stats schema according to the data filter
+        StructType statsSchema = DataSkippingUtils.getStatsSchema(metadata.getSchema());
```

Comment:
so we are converting the entire data schema into the stats schema, irrespective of which stats actually exist right? when we pass this to json parser, is the json parser creating large number of null-values column vectors for all the columns whose stats does not exist?

Reply:
+1. We should be able to get the columns used in the predicate.

Reply:
We have no way of concretely knowing ahead of time which stats exist. But the follow-up PR I'm working on prunes this schema to only include any stats columns used in the generated skipping predicate.

Reply:
if we already know the exact issue # that will be fixing/improving this, can we include that next to the `TODO`?